addons/a11y/src/components/Panel.js

```diff
@@ -1,65 +1,93 @@
 import React, { Component } from 'react';
-import addons from '@storybook/addons';
+import PropTypes from 'prop-types';
 
 import styled from 'react-emotion';
 
-import { CHECK_EVENT_ID } from '../shared';
+import { STORY_RENDERED } from '@storybook/core-events';
+import { ActionBar, ActionButton } from '@storybook/components';
+
+import { CHECK_EVENT_ID, RERUN_EVENT_ID, REQUEST_CHECK_EVENT_ID } from '../shared';
 
 import Tabs from './Tabs';
 import Report from './Report';
 
-const Passes = styled('span')({
-  color: '#0D6731',
-});
+const Passes = styled('span')(({ theme }) => ({
+  color: theme.successColor,
+}));
 
-const Violations = styled('span')({
-  color: '#AC2300',
-});
+const Violations = styled('span')(({ theme }) => ({
+  color: theme.failColor,
+}));
 
 class Panel extends Component {
-  constructor(props, ...args) {
-    super(props, ...args);
-    this.state = {
-      passes: [],
-      violations: [],
-    };
-    this.channel = addons.getChannel();
-
-    this.onUpdate = this.onUpdate.bind(this);
-  }
+  static propTypes = {
+    active: PropTypes.bool.isRequired,
+    channel: PropTypes.shape({
+      on: PropTypes.func,
+      emit: PropTypes.func,
+      removeListener: PropTypes.func,
+    }).isRequired,
+  };
+
+  state = {
+    passes: [],
+    violations: [],
+  };
 
   componentDidMount() {
-    this.channel.on(CHECK_EVENT_ID, this.onUpdate);
+    this.props.channel.on(CHECK_EVENT_ID, this.onUpdate);
+    this.props.channel.on(STORY_RENDERED, this.requestCheck);
+    this.props.channel.on(RERUN_EVENT_ID, this.requestCheck);
+  }
+
+  componentDidUpdate(prevProps) {
+    if (!prevProps.active && this.props.active) {
+      this.requestCheck();
+    }
   }
 
   componentWillUnmount() {
-    this.channel.removeListener(CHECK_EVENT_ID, this.onUpdate);
+    this.props.channel.removeListener(CHECK_EVENT_ID, this.onUpdate);
+    this.props.channel.removeListener(STORY_RENDERED, this.requestCheck);
+    this.props.channel.removeListener(RERUN_EVENT_ID, this.requestCheck);
   }
 
-  onUpdate({ passes, violations }) {
+  onUpdate = ({ passes, violations }) => {
     this.setState({
       passes,
       violations,
     });
-  }
+  };
+
+  requestCheck = () => {
+    if (this.props.active) {
+      this.props.channel.emit(REQUEST_CHECK_EVENT_ID);
+    }
+  };
 
   render() {
     const { passes, violations } = this.state;
+    const { active } = this.props;
 
-    return (
-      <Tabs
-        tabs={[
-          {
-            label: <Violations>{violations.length} Violations</Violations>,
-            panel: <Report passes={false} items={violations} empty="No a11y violations found." />,
-          },
-          {
-            label: <Passes>{passes.length} Passes</Passes>,
-            panel: <Report passes items={passes} empty="No a11y check passed" />,
-          },
-        ]}
-      />
-    );
+    return active ? (
+      <div>
+        <Tabs
+          tabs={[
```

Comment:
Implement this as `children`?

Reply:
Well I could I'm unsure it would be an improvement. As each tab gets a `active` prop passed from the `Tabs`|`TabState` component.

Do you have an implementation in mind, that looks better than this?

Reply:
This would look better to me:

```
<Tabs>
  <Tab label={<Violations>{violations.length} Violations</Violations>}>
    <Report passes={false} items={violations} empty="No a11y violations found." />
  <Tab>
  <Tab label={<Passes>{passes.length} Passes</Passes>}>
    <Report passes items={passes} empty="No a11y check passed" />
  </Tab>
</Tabs>
```

To implement it, you can just change [those lines](https://github.com/storybooks/storybook/blob/c49dff998321bdb8447cd6b19d0a30b2fb67eaf0/addons/a11y/src/components/Tabs.js#L65-L74) to

```
{React.Children.map(children (tab, index) => (
  <Item
    // eslint-disable-next-line react/no-array-index-key
    key={index}
    active={active === index ? true : undefined}
    onClick={() => this.onToggle(index)}
  >
    {tab.props.label}
  </Item>
))}
```

Reply:
Ideally, we could also use some tab identifier other than index

Reply:
just pushed a refactor to make this component behave more like your proposal. Do you like it?